kata-agent-policy: fails to compile on Darwin due to libc::S_IF* mode_t width mismatch

# Get your issue reviewed faster

1. Reproduced on `main` at [`639ff357`](https://github.com/kata-containers/kata-containers/commit/639ff3578d4d8f01627cbd79400503be814d8f8c) and later.
2. `kata-collect-data.sh` doesn't apply, this is a build time (in tests) bug on the Darwin platform (where Kata Containers isn't deployed).

# Description of problem

`libc::S_IFMT` / `S_IFREG` / `S_IFDIR` / `S_IFLNK` are `u32` on Linux but `u16` on Darwin.
In `src/agent/policy/src/policy.rs`, `impl From<u32> for FileType` and its tests mix these constants with `u32` values, so `kata-agent-policy` fails to compile on Darwin.

`kata-agent-policy` is a dev-dep of `genpolicy`, so the break propagates to `cargo test -p genpolicy`.

Introduced by [`639ff357`](https://github.com/kata-containers/kata-containers/commit/639ff3578d4d8f01627cbd79400503be814d8f8c) (CopyFile symlink restriction, [GHSA-q49m-57vm-c8cc](https://github.com/kata-containers/kata-containers/security/advisories/GHSA-q49m-57vm-c8cc)).

**Steps to reproduce** (on macOS):

```sh
git clone https://github.com/kata-containers/kata-containers
cd kata-containers/src/tools/genpolicy
cargo test -p genpolicy --lib --no-run
```

# Expected result

`genpolicy`'s test build succeeds on Darwin.

# Actual result

```
error[E0308]: mismatched types
error[E0308]: mismatched types
error[E0308]: mismatched types
error[E0308]: mismatched types
error[E0277]: no implementation for `u32 & u16`
error: could not compile `kata-agent-policy` (lib) due to 5 previous errors
```

# Further information

**Proposed fix:** widen the libc constants to `u32`.

- In `FileType::from`, use local `const S_IFMT: u32 = libc::S_IFMT as u32;` (and similar for the others) so the match keeps its table-of-patterns shape. `u32::from` would be more idiomatic but isn't `const` on stable.
- In the tests, use `u32::from(libc::S_IFREG)` instead of bare `libc::S_IFREG`, lossless on both platforms and clippy-clean.

PR to follow.

--- a/src/agent/policy/src/policy.rs
+++ b/src/agent/policy/src/policy.rs
@@ -276,11 +276,20 @@ pub enum FileType {
 }
 
 impl From<u32> for FileType {
+    // libc::S_IF* are mode_t, which is u16 on Darwin/BSD and u32 on Linux. The
+    // `as u32` cast is required for Darwin but a no-op on Linux, which trips
+    // clippy::unnecessary_cast. This is the documented libc-portability case
+    // from https://github.com/rust-lang/rust-clippy/issues/6466.
+    #[allow(clippy::unnecessary_cast)]
     fn from(raw_mode: u32) -> Self {
-        match raw_mode & libc::S_IFMT {
-            libc::S_IFREG => Self::Regular,
-            libc::S_IFDIR => Self::Directory,
-            libc::S_IFLNK => Self::Symlink,
+        const S_IFMT: u32 = libc::S_IFMT as u32;
+        const S_IFREG: u32 = libc::S_IFREG as u32;
+        const S_IFDIR: u32 = libc::S_IFDIR as u32;
+        const S_IFLNK: u32 = libc::S_IFLNK as u32;
+        match raw_mode & S_IFMT {
+            S_IFREG => Self::Regular,
+            S_IFDIR => Self::Directory,
+            S_IFLNK => Self::Symlink,
             _ => Self::Unknown,
         }
     }
@@ -338,6 +347,10 @@ impl std::convert::TryFrom<&CopyFileRequest> for PolicyCopyFileRequest {
 }
 
 #[cfg(test)]
+// libc::S_IF* constants are u16 on Darwin/BSD and u32 on Linux, and the test
+// cases below cast them to u32 to match the file_mode field type. The cast is
+// a no-op on Linux (see https://github.com/rust-lang/rust-clippy/issues/6466).
+#[allow(clippy::unnecessary_cast)]
 mod tests {
     use super::*;
     use std::convert::TryInto;
@@ -356,12 +369,12 @@ mod tests {
             TestCase {
                 name: "regular".to_owned(),
                 input: CopyFileRequest {
-                    file_mode: libc::S_IFREG,
+                    file_mode: libc::S_IFREG as u32,
                     path: "/foo/bar".to_owned(),
                     ..Default::default()
                 },
                 output: Some(PolicyCopyFileRequest {
-                    file_mode: libc::S_IFREG,
+                    file_mode: libc::S_IFREG as u32,
                     file_type: FileType::Regular,
                     path: "/foo/bar".to_owned(),
                     ..Default::default()
@@ -370,12 +383,12 @@ mod tests {
             TestCase {
                 name: "directory".to_owned(),
                 input: CopyFileRequest {
-                    file_mode: libc::S_IFDIR,
+                    file_mode: libc::S_IFDIR as u32,
                     path: "/foo".to_owned(),
                     ..Default::default()
                 },
                 output: Some(PolicyCopyFileRequest {
-                    file_mode: libc::S_IFDIR,
+                    file_mode: libc::S_IFDIR as u32,
                     file_type: FileType::Directory,
                     path: "/foo".to_owned(),
                     ..Default::default()
@@ -384,12 +397,12 @@ mod tests {
             TestCase {
                 name: "socket".to_owned(),
                 input: CopyFileRequest {
-                    file_mode: libc::S_IFSOCK,
+                    file_mode: libc::S_IFSOCK as u32,
                     path: "/foo/sock".to_owned(),
                     ..Default::default()
                 },
                 output: Some(PolicyCopyFileRequest {
-                    file_mode: libc::S_IFSOCK,
+                    file_mode: libc::S_IFSOCK as u32,
                     file_type: FileType::Unknown,
                     path: "/foo/sock".to_owned(),
                     ..Default::default()
@@ -398,12 +411,12 @@ mod tests {
             TestCase {
                 name: "mixed".to_owned(),
                 input: CopyFileRequest {
-                    file_mode: libc::S_IFDIR | libc::S_IFREG,
+                    file_mode: libc::S_IFDIR as u32 | libc::S_IFREG as u32,
                     path: "/foo/dunno".to_owned(),
                     ..Default::default()
                 },
                 output: Some(PolicyCopyFileRequest {
-                    file_mode: libc::S_IFDIR | libc::S_IFREG,
+                    file_mode: libc::S_IFDIR as u32 | libc::S_IFREG as u32,
                     file_type: FileType::Unknown,
                     path: "/foo/dunno".to_owned(),
                     ..Default::default()
@@ -412,12 +425,12 @@ mod tests {
             TestCase {
                 name: "all".to_owned(),
                 input: CopyFileRequest {
-                    file_mode: libc::S_IFMT,
+                    file_mode: libc::S_IFMT as u32,
                     path: "/wat".to_owned(),
                     ..Default::default()
                 },
                 output: Some(PolicyCopyFileRequest {
-                    file_mode: libc::S_IFMT,
+                    file_mode: libc::S_IFMT as u32,
                     file_type: FileType::Unknown,
                     path: "/wat".to_owned(),
                     ..Default::default()
@@ -441,12 +454,12 @@ mod tests {
                 name: "link/valid".to_owned(),
                 input: CopyFileRequest {
                     data: b"..data/foo".to_vec(),
-                    file_mode: libc::S_IFLNK,
+                    file_mode: libc::S_IFLNK as u32,
                     path: "/foo/lnk".to_owned(),
                     ..Default::default()
                 },
                 output: Some(PolicyCopyFileRequest {
-                    file_mode: libc::S_IFLNK,
+                    file_mode: libc::S_IFLNK as u32,
                     file_type: FileType::Symlink,
                     symlink_target: Some("..data/foo".to_owned()),
                     path: "/foo/lnk".to_owned(),
@@ -456,7 +469,7 @@ mod tests {
             TestCase {
                 name: "link/invalid".to_owned(),
                 input: CopyFileRequest {
-                    file_mode: libc::S_IFLNK,
+                    file_mode: libc::S_IFLNK as u32,
                     data: vec![0x00, 0xFF, 0xFF, 0x00],
                     ..Default::default()
                 },

